firedrake/slate/slac/kernel_builder.py

```diff
@@ -752,26 +1037,50 @@ def generate_tsfc_calls(self, terminal, loopy_tensor):
                                              within_inames=frozenset(inames_dep),
                                              predicates=predicates, id=key)
 
-                yield insn, kinfo.kernel.code
+                yield insn, {kinfo.kernel.name: kinfo.kernel.code}
+
+        # tsfc yields no kernels if they'd reduce to T0 = 0
+        if not cxt_kernels:
+            yield (None, None)
 
 
 class SlateWrapperBag(object):
 
-    def __init__(self, coeffs):
+    def __init__(self, coeffs, new_coeffs={}, name=""):
         self.coefficients = coeffs
-        self.inames = OrderedDict()
+        self.action_coefficients = new_coeffs
         self.needs_cell_orientations = False
         self.needs_cell_sizes = False
         self.needs_cell_facets = False
         self.needs_mesh_layers = False
-        self.call_name_generator = UniqueNameGenerator(forced_prefix="tsfc_kernel_call_")
+        self.num_facets = None
+        self.call_name_generator = UniqueNameGenerator()
         self.index_creator = IndexCreator()
+        self.name = name
+
+    def copy_extra_args(self, other):
+        self.coefficients = other.coefficients
+        if not self.needs_cell_orientations:
+            self.needs_cell_orientations = other.needs_cell_orientations
+        if not self.needs_cell_sizes:
+            self.needs_cell_sizes = other.needs_cell_sizes
+        if not self.needs_cell_facets:
+            self.needs_cell_facets = other.needs_cell_facets
+            self.num_facets = other.num_facets
+        if not self.needs_mesh_layers:
+            self.needs_mesh_layers = other.needs_mesh_layers
+
+    def copy_coefficients(self, coeffs=None, new_coeffs=None):
+        if coeffs:
+            self.coefficients = coeffs
+        if new_coeffs:
```

Comment:
Could this be renamed `action_coeffs` or similar? Having `coeffs` and `new_coeffs` is not clear to me.